### PR TITLE
[feature] user authentication and user specific bookshelf

### PR DESCRIPTION
**Pull request by free-soul**

### CODE REVIEW

File: client/src/pages/Login/index.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  trackChange = (name) => (event) => {
+    this.setState({ [name]: event.target.value });
+  }
+
+  render() {
+    return (
+      <div className={styles.rootContainer}>
+        <div className={styles.subContainer}>
+          <div className={styles.headingContainer}>
+              <label>LOGIN</label>
```
Comment (free-soul) [inline client/src/pages/Login/index.js]: indendation!

Comment (free-soul) [inline client/src/pages/Login/index.js]: indendation!

File: client/src/pages/Login/styles.scss [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+.rootContainer {
+  height: 100%;
+  display: flex;
+  justify-content: center;
+  background: white;
+  padding-top: 7%;
+
+  .subContainer {
+    max-width: 650px;
+    width: 41.66666667%;
```
Comment (free-soul) [inline client/src/pages/Login/styles.scss]: no vague values.

File: client/src/pages/Login/styles.scss [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        min-width: 25%;
+        border: 1px solid transparent;
+        border-radius: 4px;
+        margin: 0px 15px 0px 0px;
+      }
+    }
+  }
+
+}
+
```
Comment (free-soul) [inline client/src/pages/Login/styles.scss]: need a strategy to refactor our styles.

File: client/src/routes/index.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import App from 'src/containers/App';
+import Home from 'src/pages/Home';
+import Login from 'src/pages/Login';
+
+function requireLogin({ params }, replace) {  // eslint-disable-line
```
Comment (free-soul) [inline client/src/routes/index.js]: Migrate to [React Router 4.x](https://reacttraining.com/react-router/web/example/auth-workflow).

File: client/webpack/dev.js [hunk 5]
```diff
@@ -15,8 +15,12 @@ module.exports = {
     historyApiFallback: true,
     publicPath: DEV_URL,
     port: PORT,
+    // [TODO] Disabling proxy server for now; enable this once we are ready to hit APIs
     proxy: {
-      '/': 'http://localhost:4001' // Proxying to dev-api running at port 4001
+      "/api": {
```
Comment (free-soul) [inline client/webpack/dev.js]: we use single-quotes.

File: server/index.js [hunk 6]
```diff
@@ -1,43 +1,70 @@
...
-app.listen(config.apiPort, function(err){
-	// incase of an error return it
-	if (err) throw err;
-	// keep the user aware of the status of the server
-	console.log("Server running on port:", config.apiPort);
+app.listen(config.API_PORT, (err) => {
+  // incase of an error return it
+  if (err) throw err;
+  // keep the user aware of the status of the server
+  console.log('Server running on port: ', config.API_PORT); // eslint-disable-line
```
Comment (free-soul) [inline server/index.js]: colors FTW!

File: server/src/api/book/index.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+const codes = require('../../config/codes');
+const midWare = require('../middlewares');
+// Schemas
+const Book = require('../../models/book');
+// Thunks
+const bookThunks = require('./thunks');
+const userThunks = require('../user/thunks');
+
+// GET all books
+router.get('/all_books', (req, res, next) => {
```
Comment (free-soul) [inline server/src/api/book/index.js]: think of a better endpoint name.

File: server/src/api/book/index.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      {
+        reqBody: req.body,
+        reqSession: req.session,
+        bookId: newBook._id,
+        overwrite: true
+      },
+      (bookId, newUserBook) => {
+        res.status(200).send({
+          message: codes.BOOK.S_UPD,
+          result: Object.assign({}, newBook.toJSON(), newUserBook)
```
Comment (free-soul) [inline server/src/api/book/index.js]: `newBook.toJSON()` this is repeating! find the root cause.

File: server/src/api/book/thunks.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  Book.findOne(
+    {
+      $or: [
+        { _id: reqBody._id },
+        {
+          $and: [
+            { title: reqBody.title },
+            { author: reqBody.author }
+          ]
+        }
```
Comment (free-soul) [inline server/src/api/book/thunks.js]: is there a better way to write this?

File: client/src/pages/Login/index.js (hunk 1 above)
Comment (jainkjohnson) [inline client/src/pages/Login/index.js]: Done

File: client/src/pages/Login/styles.scss (hunk 2 above)
Comment (jainkjohnson) [inline client/src/pages/Login/styles.scss]: Done

File: client/webpack/dev.js (hunk 5 above)
Comment (jainkjohnson) [inline client/webpack/dev.js]: Done

File: client/src/pages/Login/index.js [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+  getChildContext() {
+    return {
+      reactIconBase: {
+        size: appConfig.ICON_SIZE,
+      },
+    };
+  }
+
+  componentDidMount() {
```
Comment (jainkjohnson) [inline client/src/pages/Login/index.js]: empty componentDidMount!

File: client/src/pages/Login/index.js [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+  };
+
+  static childContextTypes = {
+    reactIconBase: object,
+  };
+
+  static defaultProps = {
+  };
+
+  state = {
```
Comment (jainkjohnson) [inline client/src/pages/Login/index.js]: states are not added

File: server/src/api/book/thunks.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+const Book = require('../../models/book');
+
+/**
+ * A callback function that should be executed if everything goes right.
+ * @callback onSuccessCallback
+ * @param {*}
+ */
+
+/**
+ * A callback function that should be executed if somethinh goes wrong.
```
Comment (jainkjohnson) [inline server/src/api/book/thunks.js]: spelling of something

File: server/src/api/user/thunks.js [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+const utils = require('../../utils');
+
+/**
+ * A callback function that should be executed if everything goes right.
+ * @callback onSuccessCallback
+ * @param {*}
+ */
+
+/**
+ * A callback function that should be executed if somethinh goes wrong.
```
Comment (jainkjohnson) [inline server/src/api/user/thunks.js]: spelling

**Pull request merged**